File: generate.ts
import fetch from 'cross-fetch';
import fs from 'fs/promises';
import { JSDOM } from 'jsdom';

async function fetchProblem(exercise: string) {
    const result = await (await fetch(`https://rosalind.info/problems/${exercise}`)).text();
    const dom = new JSDOM(result);
    const sampleDataset = dom.window.document
        .querySelector('#sample-dataset + .codehilite>pre')
        ?.innerHTML.trim()
        .split(/\s/);
    const sampleOutput = dom.window.document
        .querySelector('#sample-output + .codehilite>pre')
        ?.innerHTML.trim()
        .split(/\s/)
        .map((d) => (d.match(/^\d+$/) ? +d : d));
    return [sampleDataset, sampleOutput];
}

async function writeImpl(exercise: string, chapter: string) {
    const dir = `./exercises/ch${chapter}`;
    await fs.mkdir(dir, { recursive: true });

    await fs.writeFile(
        `${dir}/${exercise}.ts`,
        `function foobar(pattern: string): string {
    return pattern
}

export default function ${exercise}(params: string[]) {
    const [pattern] = params;
    return foobar(pattern);
}
        `,
    );
}

async function writeTest(
    exercise: string,
    chapter: string,
    sampleDataset?: (string | number)[],
    sampleOutput?: (string | number)[],
) {
    const dir = `./exercises/ch${chapter}`;
    await fs.mkdir(dir, { recursive: true });

    await fs.writeFile(
        `${dir}/${exercise}.test.ts`,
        `import ${exercise} from './${exercise}';

test('example correct', () => {
    expect(${exercise}(${JSON.stringify(sampleDataset)})).toEqual(${JSON.stringify(sampleOutput)});
});
        `,
    );
}

async function generate() {
    if (process.argv.length !== 3) throw new Error('Must provide an exercise!');

    const [exercise, chapter] = process.argv[2].toLowerCase().match(/ba(?<chapter>\d+)[a-z]/) ?? [];
    if (!exercise || !chapter) throw new Error('Must provide a valid exercise!');

    console.info(`Generating exercise ${exercise} from chapter ${chapter}...`);

    const [sampleDataset, sampleOutput] = await fetchProblem(exercise);

    await Promise.all([writeImpl(exercise, chapter), writeTest(exercise, chapter, sampleDataset, sampleOutput)]);
}

generate();

File: generate.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"strings"

	"github.com/PuerkitoBio/goquery"
)

const implTempl = `
package exercises

func foobar(a []byte, b []byte) int {
	return 1
}

func (ex Exercise) %s(params [][]byte) int {
	a := params[0]
	b := params[1]

	result := foobar(a, b)

	return result
}
`

const testTempl = `
package exercises_test

import (
	"testing"

	"github.com/vgalaktionov/bioinformatics/exercises"
)

func Test%s(t *testing.T) {
	expected := "%s"
	result := exercises.Exercise{}.%s([][]byte{%s})

	if result != expected {
		t.Errorf("expected: %%d; got: %%d", expected, result)
	}
}

func Benchmark%s(b *testing.B) {
	for i := 0; i < b.N; i++ {
		exercises.Exercise{}.%s([][]byte{%s})
	}
}
`

func Generate() {
	ex := os.Args[2]
	exUpper := strings.ToUpper(ex)
	log.Printf("generating exercise: %s\n", ex)
	// Request the HTML page.
	res, err := http.Get(fmt.Sprintf("https://rosalind.info/problems/%s", ex))
	if err != nil {
		log.Fatal(err)
	}
	defer res.Body.Close()
	if res.StatusCode != 200 {
		log.Fatalf("status code error: %d %s", res.StatusCode, res.Status)
	}

	// Load the HTML document
	doc, err := goquery.NewDocumentFromReader(res.Body)
	if err != nil {
		log.Fatal(err)
	}

	// query for test dataset
	var dataset string
	doc.Find("#sample-dataset + .codehilite>pre").Each(func(i int, s *goquery.Selection) {
		dataset += s.First().Text()
	})

	// query for return dataset
	var output string
	doc.Find("#sample-output + .codehilite>pre").Each(func(i int, s *goquery.Selection) {
		output += s.First().Text()
	})
	output = strings.Trim(output, " \n")

	// format inputs for generated testcase
	datasetInputs := []string{}
	for _, p := range strings.Fields(dataset) {
		datasetInputs = append(datasetInputs, fmt.Sprintf(`[]byte("%s")`, p))
	}
	joined := strings.Join(datasetInputs, ", ")

	// write files to disk
	ioutil.WriteFile(fmt.Sprintf("./exercises/%s.go", ex), []byte(fmt.Sprintf(implTempl, strings.ToUpper(ex))), 0777)

	ioutil.WriteFile(fmt.Sprintf("./exercises/%s_test.go", ex), []byte(fmt.Sprintf(testTempl, exUpper, output, exUpper, joined, exUpper, exUpper, joined)), 0777)
}
